File: Python/Becks4Eq.py
# -*- coding: utf-8 -*-
"""
Solve Beck's system of ODEs
Created on Wed Apr 29 15:12:24 2015

@author: Christopher Strickland
"""
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import time
from __future__ import division

##### Variables go here #####
N0 = 4.5e-5
tpts = np.concatenate((np.array([0]),np.linspace(7000,7500,501)))
x0 = np.array([2.1e-7,4.56e-7,4.3e-5,4.5e-5])

#growth rates (1/sec)
#values below are in 1/day; divide by 24*3600 to get seconds
munr = 12.0
munc = 6.0
mupr =2.2
mupc = 2.2/8.0

#mass (g)
mr = 1.6e-12
mc = 8.2e-12
mp = 9.8e-10

#half-saturation constants (g/cc)
knr = 8.0e-6
knc = 8.0e-6
kpr = 1.0e-6
kpc = 1.0e-6

#death rates (1/sec)
#these values are in 1/day - divide by 24*3600 to get seconds
dr = 0.5/10.0
dc = 0.25/100.0
dp = 0.08/100.0

#yield coefficients
ypr = 0.12
ypc = 0.12
ynr = 0.1
ync = 0.1    

#ODE function
def becks4Eq(x,t,N0,D):
    dx = np.zeros(4)
    
    #ODEs
    dx[0] = x[0]*(munr*x[3]/(x[3]+knr)-dr) - \
    mupr/ypr*mp/mr*x[0]/(kpr/mr+x[0])*x[2] - D*x[0]
    
    dx[1] = x[1]*(munc*x[3]/(x[3]+knc)-dc) - \
    mupc/ypc*mp/mc*x[1]/(kpc/mc+x[1])*x[2] - D*x[1]
    
    dx[2] = x[2]*(mupr*x[0]/(kpr/mr+x[0]) + \
    mupc*x[1]/(kpc/mc+x[1])-dp) - D*x[2]
    
    dx[3] = D*N0 - x[0]*munr*mr/ynr*x[3]/(knr+x[3]) - \
    x[1]*munc*mc/ync*x[3]/(knc+x[3]) - D*x[3]
    
    return dx

##### Solve procedure goes here #####
for D in xrange(0.05,2.01,0.01):
    Usol = sp.integrate.odeint(becks4Eq,x0,tpts,(N0,D))
    plt.figure(1)
    plt.rc('text', usetex=True)
    plt.hold(True)
    plt.plot(D*np.ones(len(tpts)-1),Usol[1:,0],marker='.',markersize=2)
    plt.title(r"Bifurcation Diagram for $R$ vs $D$")
    plt.xlabel(r"Values for $D$")
    plt.ylabel(r"Rod Species $R$")
    plt.draw()
    
    plt.figure(2)
    plt.rc('text', usetex=True)
    plt.hold(True)
    plt.plot(D*np.ones(len(tpts)-1),Usol[1:,1],marker='.',markersize=2)
    plt.title(r"Bifurcation Diagram for $C$ vs $D$")
    plt.xlabel(r"Values for $D$")
    plt.ylabel(r"Cocci Species $C$")
    plt.draw()
    
    print(D)
    plt.sleep(0.05)
    
    

File: MATLAB/becks4Eq.m
function xp = becks4Eq(t,x)

xp = zeros(4,1);

global N0
global D

% growth rates (1/sec)
% values below are in 1/day; divide by 24*3600 to get seconds
munr = 12.0 ;  % original
% munr = 12.157197177976418 ;  % original
% munr = 6 ;
munc = 6.0 ;   % original
% munc = 2.5009944119115524 ;
% mupr = 1.6919353004300042 ;
% mupc = 0.13790519501669446 ;
mupr = 2.2 ;  % original
mupc = 2.2/8.0 ;  % original
% mupc = 2.2 ;
%
% mass (g)
mr = 1.6e-12 ;
mc = 8.2e-12 ;
mp = 9.8e-10 ;
%
% half-saturation constants (g/cc)
knr = 8.0e-6 ;
knc = 8.0e-6 ;
kpr = 1.0e-6 ;
kpc = 1.0e-6 ;
%
% death rates (1/sec)
% values below are in 1/day; divide by 24*3600 to get seconds
dr = 0.5/10; % original
% dr = 0.03535901442365246;
dc = 0.25/100; % original
% dc = 0.002636146332980288
% dr = 0.5 ;
% dc = 0.25/10 ;
dp = 0.08/100 ; % original
% dp = 4.96579147003975e-4;
%
% initial nutrient (g/cc)
% global d;
% N0 = 2.3e-4 ; % original
% N0 = 2.3e-4 ;
% N0 = d;
%
% yield coefficients
ypr = 0.12 ;
ypc = 0.12 ;
ynr = 0.1 ;
ync = 0.1 ;
%
% dilution rate
% values below are in 1/day; divide by 24*3600 to get seconds
% D = 0.14512023872492155;
% D = 0.14660367473913746;
% D = 0.1451;  % 
% D = 0.6 ;  % original
%
%
% Differential equations

xp(1,1) = x(1,1)*(munr*x(4,1)/(x(4,1)+knr) - dr) - mupr/ypr*mp/mr*x(1,1)/(kpr/mr+x(1,1))*x(3,1) - D*x(1,1) ;

xp(2,1) = x(2,1)*(munc*x(4,1)/(x(4,1)+knc) - dc) - mupc/ypc*mp/mc*x(2,1)/(kpc/mc+x(2,1))*x(3,1) - D*x(2,1) ;

xp(3,1) = x(3,1)*(mupr*x(1,1)/(kpr/mr+x(1,1)) + mupc*x(2,1)/(kpc/mc+x(2,1)) - dp) - D*x(3,1) ;

xp(4,1) = D*N0 - x(1,1)*munr*mr/ynr*x(4,1)/(knr+x(4,1)) - x(2,1)*munc*mc/ync*x(4,1)/(knc+x(4,1)) - D*x(4,1);


